API for adding and removing table constraints

kgodey opened the pull request:
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, open one before creating this pull request. -->
Fixes #326

<!-- Concisely describe what the pull request does. -->
This pull request adds an API to add and remove constraints from tables.

## Technical details
<!-- Add any other information or technical details about the implementation; or delete the section entirely. -->
Per the design specs for this milestone, we will only support:
- listing constraints
- adding a unique constraint
- deleting a constraint

Adding foreign keys and primary keys will be supported in:
- #440 
- #451 

Support for adding check constraints is not planned for the MVP.

## Refactors
There's some miscellaneous cleanup and refactoring in here.

## TODO
- [x] DB function for adding a unique constraint
- [x] DB function for dropping a constraint
- [x] DB tests for adding a unique constraint
- [x] DB tests for dropping a constraint
- [x] API to read constraints
- [x] API to add a unique constraint
- [x] API to drop a constraint
- [x] Tests for API to list constraints
- [x] Test for API to retrieve a constraint
- [x] Tests for API to add a unique constraint
- [x] Tests for API to drop a constraint
- [x] Test for trying to set a unique constraint for a column with non-unique values
- [x] Test for trying to duplicate constraint name
- [x] Ensure constraint names are idiomatic
- [x] Ensure constraint names are unique
- [x] Add constraints URL to table serializer
- [x] Add constraint ID to serailizer

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. --> 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [x] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

eito-fis commented on mathesar/views/api.py:
```diff
@@ -290,3 +292,61 @@ def create(self, request):
+
+    def destroy(self, request, pk=None, table_pk=None):
+        table = Table.objects.get(id=table_pk)
+        table.drop_constraint(pk)
```
Could catch a 404 here?

kgodey replied:
Good catch, I'll fix it.

eito-fis reviewed (approved):
Looks good to me! Just a small 404 nitpick.

kgodey commented:
@eito-fis I added handling of 404 errors for deletion and also for every instance where we're trying to get a table by `pk`. I added tests for both scenarios.

mathemancer reviewed (commented):
Since all constraints have OIDs in the Postgres, I think we should have a model for them to hold that mapping for referring to them (similar to tables and schemata), both for consistency with those objects, and convenience.  This would let us notice whether a constraint was dropped or renamed, and also give us more stability when looking up and working with constraints.  

Documentation of the relevant system catalog table:  https://www.postgresql.org/docs/13/catalog-pg-constraint.html

kgodey commented:
> Since all constraints have OIDs in the Postgres, I think we should have a model for them to hold that mapping for referring to them (similar to tables and schemata), both for consistency with those objects, and convenience. This would let us notice whether a constraint was dropped or renamed, and also give us more stability when looking up and working with constraints.
> 
> Documentation of the relevant system catalog table: https://www.postgresql.org/docs/13/catalog-pg-constraint.html

That makes sense, I'll refactor. I hadn't realized that constraints had OIDs, it definitely makes sense to store them in a model.

powellc reviewed (approved):
Excellent work, @kgodey !  I need to get more familiar with SQLAlchemy to be confident reviewing that code, but the DRF and general python code looks great. Just a few comments but none of them would block a merge. 🚀

kgodey commented on mathesar/signals.py:
```diff
@@ -0,0 +1,22 @@
+
+
+@receiver(post_save, sender=Table)
+def create_new_constraints(sender, instance, created, **kwargs):
```
I ended up removing the `pre_delete` signal because I realized that we delete constraint models in other places and we don't always automatically want to drop them. However, I did rename the function to be clearer and added a comment.

kgodey commented on mathesar/reflection.py:
```diff
@@ -39,33 +40,56 @@ def reflect_databases():
+                table=models.Table.current_objects.get(oid=db_constraint['conrelid'])
+            )
+        except models.Table.DoesNotExist:
+            pass
```
I agree the readability could be improved. Let me know if the version I pushed looks better. There really shouldn't be duplicate OIDs, given that it's the primary identifier of the table in the database, so I deliberately didn't handle that case, I'd rather it throw an error.

mathemancer replied:
The OIDs do roll over at some point, but that would be an _extreme_ edge case, since the user would have needed to create many _many_ more tables than the throughput of our app is probably going to enable in a reasonable timeframe.  I definitely think throwing an error would appropriate.  An amusing (if ancient) thread from the PostgreSQL mailing list:

https://www.postgresql.org/message-id/Pine.GSO.4.02A.10004101348060.8269-100000%40Falk.DoCS.UU.SE

powellc replied:
@kgodey looks great, thank you!

kgodey commented:
Thanks for the review, @powellc!

eito-fis commented on mathesar/views/api.py:
```diff
@@ -292,3 +301,50 @@ def create(self, request):
         return create_datafile(request, serializer.validated_data)
+
+
+class ConstraintViewSet(viewsets.GenericViewSet, ListModelMixin, RetrieveModelMixin, CreateModelMixin, DestroyModelMixin):
```
Do we need `CreateModelMixin` and `DestroyModelMixin` if we overwrite those methods?

kgodey replied:
No, we don't. :)

eito-fis reviewed (approved):
Some tiny things, but otherwise LGTM :+1: 

Also, unrelated to this PR, but it seems like we're missing 404s for a lot of the records endpoint and most of the pagination classes. Worth opening an issue for?

kgodey commented on db/constraints.py:
```diff
@@ -0,0 +1,108 @@
+    with warnings.catch_warnings():
+        warnings.filterwarnings("ignore", message="Did not recognize type")
+        pg_constraint = Table("pg_constraint", metadata, autoload_with=engine)
+        # We only want to select constraints attached to a table.
```
It is not, good find, I copied it from the other method and didn't update it.

kgodey commented on mathesar/models.py:
```diff
@@ -211,6 +215,66 @@ def update_record(self, id_value, record_data):
+        return None
+
+    def add_constraint(self, constraint_data):
+        if 'type' not in constraint_data or constraint_data['type'] != constraints.ConstraintType.UNIQUE.value:
```
Good catch, agreed.

kgodey commented:
@eito-fis Thanks for the review! I think it makes sense to create an issue for the missing 404s. Would you be able to do that?

mathemancer commented on mathesar/views/api.py:
```diff
@@ -292,3 +301,50 @@ def create(self, request):
+        except ProgrammingError as e:
+            if type(e.orig) == DuplicateTable:
+                raise ValidationError(
+                    'Constraint with the same name already exists'
```
This will also happen if a table with the desired name already exists (or any Relation).  I suggest just using the generic "Relation" to avoid confusion.

mathemancer commented on mathesar/views/api.py:
```diff
@@ -292,3 +301,50 @@ def create(self, request):
+        serializer.is_valid(raise_exception=True)
+        # If we don't do this, the request.data QueryDict will only return the last column's name
+        # if there are multiple columns.
+        data = request.data.dict()
+        data['columns'] = request.data.getlist('columns')
```
When I try to submit POST requests via the DRF-created web ui, `request.data` seems to come through as a `dict` type here, which breaks these lines.  Am I doing something wrong?  The request I submitted was:
```
{
    "name": "aconstraint",
    "type": "unique",
    "columns": ["Status"]
}
```
The error I got was 
![image](https://user-images.githubusercontent.com/6829750/127875647-7aaeb9d7-d795-4f48-8551-bfb7bcf727ee.png)

Fixing that error just let the next line cause the error (since there's no `getlist` method of python `dict` either).

kgodey replied:
This should be fixed now. I don't know why it's a `QueryDict` sometimes and a dict other times, but both cases are handled.

mathemancer commented on mathesar/models.py:
```diff
@@ -211,6 +215,66 @@ def update_record(self, id_value, record_data):
+            constraint_data['columns']
+        )
+        engine = self.schema.database._sa_engine
+        constraint_oid = constraints.get_constraint_oid_by_name_and_table_oid(sa_constraint.name, self.oid, engine)
+        return Constraint.objects.create(oid=constraint_oid, table=self)
```
You can have more than one constraint on the same columns of the same type.  This might happen through migrations, user inattention, etc.  This could end up reflecting the wrong one, and running into confusion when trying to add a Constraint object with a previously-reflected OID, rather than the one intended.

kgodey replied:
This is fixed. Since we know the pattern of the name, I generated the name instead of trying to get the constraint by type/columns. The name should be unique.

mathemancer reviewed (changes requested):
Overall, I think the new way of doing this works, and will extend well.  There's a bug when I try to submit `POST` requests (or I'm screwing something up); see the specific comment and screenshot.

mathemancer commented on mathesar/views/api.py:
```diff
@@ -350,3 +359,52 @@ def create(self, request):
+            data = request.data.dict()
+            data['columns'] = request.data.getlist('columns')
+        try:
+            name = data['name'] if 'name' in data else None
```
I don't think `data` is defined here.

kgodey replied:
Ugh, sorry. It's fixed now, I verified through the HTML interface.

mathemancer reviewed (changes requested):
When submitting the request
```json
{
    "name": "supernewtable_unique2",
    "type": "unique",
    "columns": [
        "mathesar_id"
    ]
}
```
I get
```
web_1  | Internal Server Error: /api/v0/tables/1/constraints/
web_1  | Traceback (most recent call last):
web_1  |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 47, in inner
web_1  |     response = get_response(request)
web_1  |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 181, in _get_response
web_1  |     response = wrapped_callback(request, *callback_args, **callback_kwargs)
web_1  |   File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
web_1  |     return view_func(*args, **kwargs)
web_1  |   File "/usr/local/lib/python3.9/site-packages/rest_framework/viewsets.py", line 125, in view
web_1  |     return self.dispatch(request, *args, **kwargs)
web_1  |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
web_1  |     response = self.handle_exception(exc)
web_1  |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
web_1  |     self.raise_uncaught_exception(exc)
web_1  |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
web_1  |     raise exc
web_1  |   File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
web_1  |     response = handler(request, *args, **kwargs)
web_1  |   File "/code/mathesar/views/api.py", line 381, in create
web_1  |     name = data['name'] if 'name' in data else None
web_1  | UnboundLocalError: local variable 'data' referenced before assignment
web_1  | [04/Aug/2021 09:50:05] "POST /api/v0/tables/1/constraints/ HTTP/1.1" 500 86861
```

pavish reviewed (changes requested):
Please revert the changes in package-lock.json

The audit fixes are done in https://github.com/centerofci/mathesar/pull/508/commits/81101df0c11bb052a827754e6545fd9acb2a768b.

pavish reviewed (commented):
The changes for package-lock.json in 5ff56c0 look good to me.

mathemancer reviewed (approved):
I think it's looking good, and it works on my machine.  Thank you!
